python/orca/tutorial/pytorch/NCF/train_xshards.py

```diff
@@ -0,0 +1,210 @@
+
+import torch
+import torch.nn as nn
+import torch.optim as optim
+import torch.utils.data as data
+import torch.nn.functional as F 
+from model import NCF
+
+#Step 0: Parameters And Configuration
+
+parser = argparse.ArgumentParser()
+parser.add_argument("--dataset", 
+    type=str, 
+    default="ml-1m", 
+    help="dataset name")
+parser.add_argument("--model", 
+    type=str, 
+    default="NeuMF-end", 
+    help="model name")
+parser.add_argument("--main_path", 
+    type=str, 
+    default="./NCF-Data/", 
+    help="main path")
+parser.add_argument("--model_path", 
+    type=str, 
+    default="./models/", 
+    help="model path")
+parser.add_argument("--out", 
+    type=bool, 
+    default=True, 
+    help="save model or not")
+parser.add_argument("--cluster_mode", 
+    type=str, 
+    default="local", 
+    help="")
+parser.add_argument("--lr", 
+    type=float, 
+    default=0.001, 
+    help="learning rate")
+parser.add_argument("--dropout", 
+    type=float, 
+    default=0.0, 
+    help="dropout rate")
+parser.add_argument("--batch_size", 
+    type=int, 
+    default=256, 
+    help="batch size for training")
+parser.add_argument("--epochs", 
+    type=int, 
+    default=20, 
+    help="training epoches")
+parser.add_argument("--top_k", 
+    type=int, 
+    default=10, 
+    help="compute metrics@top_k")
+parser.add_argument("--factor_num", 
+    type=int, 
+    default=32, 
+    help="predictive factors numbers in the model")
+parser.add_argument("--num_layers", 
+    type=int, 
+    default=3, 
+    help="number of layers in MLP model")
+parser.add_argument("--num_ng", 
+    type=int, 
+    default=4, 
+    help="sample negative items for training")
+parser.add_argument("--test_num_ng", 
+    type=int, 
+    default=0, 
+    help="sample part of negative items for testing")
+parser.add_argument("--backend", 
+    type=str, 
+    default="spark", 
+    help="backend used in estimator, ray or spark are supported")
+parser.add_argument("--user_num", 
+    type=int, 
+    default=6040, 
+    help="total user num in the ml-1m dataset")
+parser.add_argument("--item_num", 
+    type=int, 
+    default=3952, 
+    help="total item num in the ml-1m dataset")
+args = parser.parse_args()
+
+#Step 1: Init Orca Context
+
+from bigdl.orca import init_orca_context, stop_orca_context, OrcaContext
+init_orca_context(cores=1, memory="8g")
+
+#Step 2: Define Dataset
+
+from bigdl.orca.data import XShards
+from bigdl.orca.data.pandas import read_csv
+from sklearn.model_selection import train_test_split
+
+def preprocess_data():   
+    data_X = read_csv(
+        args.dataset+"/ratings.dat", 
+        sep="::", header=None, names=['user', 'item'], 
+        usecols=[0, 1], dtype={0: np.int32, 1: np.int32})  
+    data_X = data_X.partition_by("user",5)#num_partitions=5 
+    return data_X
+
+# prepare the train and test datasets
+data_X = preprocess_data()
+
+# construct the train and test xshards
+def transform_to_dict(data):

```

Comment:
I think we'd better split the logic of negative sampling and transform to dict into two transform steps